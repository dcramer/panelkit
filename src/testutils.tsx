import React from "react";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { Entity } from "./types";
import HomeAssistant, { SuggestedChanges } from "./hass";

export const URL = "http://localhost:8123";

export const ACCESS_TOKEN = "example.access.token";

export const DEFAULT_CONFIG = {
  url: URL,
  accessToken: ACCESS_TOKEN,
  tiles: [],
};

export const simulateTouch = (container: any) => {
  userEvent.click(container);
};

export const getTileProps = (entity: Entity) => {
  const hass = new HomeAssistant(DEFAULT_CONFIG);

  hass.callService = jest
    .fn()
    .mockImplementation(
      (
        domain: string,
        service: string,
        serviceData: any,
        suggestedChanges: SuggestedChanges | null = null
      ) => {
        return Promise.resolve();
      }
    );
  hass.sendMessage = jest.fn().mockImplementation(() => {
    return Promise.resolve();
  });

  hass.populateEntityCache([entity]);

  return {
    entityId: entity.entity_id,
    hass,
    cameraList: [],
  };
};

export const renderTile = (
  Component: any,
  entity: Entity,
  props: { [key: string]: any } = {}
) => {
  const tileProps = {
    ...getTileProps(entity),
    ...props,
  };

  const result = render(<Component {...tileProps} />);

  return {
    ...result,
    hass: tileProps.hass,
    container: result.container.firstChild,
  };
};
