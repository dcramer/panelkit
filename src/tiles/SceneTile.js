import PropTypes from "prop-types";

import Tile, { TileProps } from "./Tile";

export default class SceneTile extends Tile {
  static propTypes = {
    ...TileProps,
    entityId: PropTypes.string.isRequired,
  };

  getWatchedEntityIds() {
    return [this.props.entityId];
  }

  onClick = async () => {
    await this.callService("scene", "turn_on", {
      entity_id: this.props.entityId,
    });
  };

  renderStatus() {}
}
