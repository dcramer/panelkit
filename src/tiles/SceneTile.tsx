import Tile, { TileProps } from "./Tile";

type SceneTileProps = TileProps & {
  entityId: string;
};

export default class SceneTile extends Tile<SceneTileProps> {
  onTouch = () => {
    this.callService("scene", "turn_on", {
      entity_id: this.props.entityId,
    });
  };

  renderStatus() {}
}
