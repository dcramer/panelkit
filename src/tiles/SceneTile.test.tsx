import { waitFor } from "@testing-library/react";

import { renderTile, simulateTouch } from "../testutils";
import SceneTile from "./SceneTile";

test("renders", async () => {
  const { container, getByText } = renderTile(SceneTile, {
    entity_id: "scene.test",
    state: "off",
    attributes: {
      friendly_name: "Test Scene",
    },
  });

  expect(getByText(/Test Scene/i)).toBeInTheDocument();
  expect(container).toMatchSnapshot();
});

test("activates on touch event", async () => {
  const { container, hass } = renderTile(SceneTile, {
    entity_id: "scene.test",
    state: "off",
    attributes: {
      friendly_name: "Test Scene",
    },
  });

  simulateTouch(container);

  await waitFor(() => {
    expect(hass.callService).toHaveBeenCalledTimes(1);
    expect(hass.callService).toHaveBeenCalledWith(
      "scene",
      "turn_on",
      {
        entity_id: "scene.test",
      },
      undefined
    );
  });
});
