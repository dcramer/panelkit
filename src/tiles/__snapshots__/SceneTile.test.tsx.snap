// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div
  class="tile state-off touchable"
  data-testid="tile"
  style="cursor: pointer;"
>
  <div
    class="tile-container"
  >
    <div
      class="tile-body"
    />
    <div
      class="tile-title"
    >
      Test Scene
    </div>
  </div>
</div>
`;
